File: API/3801-3900/3801_OrderStabilization/PY/order_stabilization_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan, Math
from StockSharp.Messages import DataType, CandleStates, Sides
from StockSharp.Algo.Strategies import Strategy


class order_stabilization_strategy(Strategy):
    def __init__(self):
        super(order_stabilization_strategy, self).__init__()

        self._order_volume = self.Param("OrderVolume", 0.1).SetGreaterThanZero()
        self._order_distance = self.Param("OrderDistancePoints", 20.0).SetGreaterThanZero()
        self._profit_threshold = self.Param("ProfitThreshold", -2.0)
        self._absolute_fixation = self.Param("AbsoluteFixation", 30.0).SetNotNegative()
        self._stabilization_points = self.Param("StabilizationPoints", 25.0).SetGreaterThanZero()
        self._expiration_minutes = self.Param("ExpirationMinutes", 20).SetNotNegative()
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(5)))

        self._buy_stop = None
        self._sell_stop = None
        self._created_at = None
        self._entry_price = 0.0
        self._previous_body = 0.0
        self._has_previous_body = False

    def GetWorkingSecurities(self):
        return [(self.Security, self._candle_type.Value)]

    def OnReseted(self):
        super(order_stabilization_strategy, self).OnReseted()
        self._clear_pending()
        self._entry_price = 0.0
        self._previous_body = 0.0
        self._has_previous_body = False

    def OnStarted2(self, time):
        super(order_stabilization_strategy, self).OnStarted2(time)
        self.SubscribeCandles(self._candle_type.Value).Bind(self._process_candle).Start()

    def _process_candle(self, candle):
        if candle.State != CandleStates.Finished:
            return

        point = float(self.Security.PriceStep) if self.Security is not None and self.Security.PriceStep is not None else 1.0
        if point <= 0:
            point = 1.0

        body = abs(float(candle.ClosePrice) - float(candle.OpenPrice))
        body_limit = float(self._stabilization_points.Value) * point

        if self.Position != 0:
            pnl = self._floating_pnl(float(candle.ClosePrice), point)
            one_small = body <= body_limit
            two_small = one_small and self._has_previous_body and self._previous_body <= body_limit

            if ((one_small and pnl > float(self._profit_threshold.Value)) or two_small or
                    (float(self._absolute_fixation.Value) > 0 and pnl >= float(self._absolute_fixation.Value))):
                self._flatten()
                self._clear_pending()
                self._previous_body = body
                self._has_previous_body = True
                return

        if (self._created_at is not None and int(self._expiration_minutes.Value) > 0 and
                (candle.OpenTime - self._created_at).TotalMinutes >= int(self._expiration_minutes.Value)):
            self._create_pending(float(candle.ClosePrice), candle.OpenTime, point)

        if self._buy_stop is None and self._sell_stop is None:
            self._create_pending(float(candle.ClosePrice), candle.OpenTime, point)

        hit_buy = self._buy_stop is not None and float(candle.HighPrice) >= self._buy_stop
        hit_sell = self._sell_stop is not None and float(candle.LowPrice) <= self._sell_stop

        if hit_buy or hit_sell:
            if hit_buy and hit_sell:
                side = Sides.Buy if float(candle.ClosePrice) >= float(candle.OpenPrice) else Sides.Sell
            else:
                side = Sides.Buy if hit_buy else Sides.Sell

            trigger = self._buy_stop if side == Sides.Buy else self._sell_stop
            volume = float(self._order_volume.Value)

            if side == Sides.Buy:
                self.BuyMarket(volume)
                self._buy_stop = None
            else:
                self.SellMarket(volume)
                self._sell_stop = None

            self._entry_price = trigger

        self._previous_body = body
        self._has_previous_body = True

    def _create_pending(self, center, time, point):
        distance = float(self._order_distance.Value) * point
        self._buy_stop = center + distance
        self._sell_stop = center - distance
        self._created_at = time

    def _floating_pnl(self, price, point):
        if self.Position == 0 or self._entry_price == 0:
            return 0.0
        direction = 1.0 if self.Position > 0 else -1.0
        move = (price - self._entry_price) * direction
        step_price = float(self.Security.StepPrice) if self.Security is not None and self.Security.StepPrice is not None else 0.0
        multiplier = float(self.Security.Multiplier) if self.Security is not None and self.Security.Multiplier is not None else 1.0
        return move / point * step_price * abs(float(self.Position)) if step_price > 0 else move * multiplier * abs(float(self.Position))

    def _flatten(self):
        if self.Position > 0:
            self.SellMarket(Math.Abs(self.Position))
        elif self.Position < 0:
            self.BuyMarket(Math.Abs(self.Position))
        self._entry_price = 0.0

    def _clear_pending(self):
        self._buy_stop = None
        self._sell_stop = None
        self._created_at = None

    def CreateClone(self):
        return order_stabilization_strategy()

File: API/3801-3900/3801_OrderStabilization/CS/OrderStabilizationStrategy.cs
using System;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Paired breakout triggers with stabilization-based exits.
/// </summary>
public class OrderStabilizationStrategy : Strategy
{
	private readonly StrategyParam<decimal> _orderVolume;
	private readonly StrategyParam<decimal> _orderDistancePoints;
	private readonly StrategyParam<decimal> _profitThreshold;
	private readonly StrategyParam<decimal> _absoluteFixation;
	private readonly StrategyParam<decimal> _stabilizationPoints;
	private readonly StrategyParam<int> _expirationMinutes;
	private readonly StrategyParam<DataType> _candleType;

	private decimal? _buyStop;
	private decimal? _sellStop;
	private DateTimeOffset? _createdAt;
	private decimal _entryPrice;
	private decimal _previousBody;
	private bool _hasPreviousBody;
	private DateTimeOffset? _entryCandleTime;

	public decimal OrderVolume { get => _orderVolume.Value; set => _orderVolume.Value = value; }
	public decimal OrderDistancePoints { get => _orderDistancePoints.Value; set => _orderDistancePoints.Value = value; }
	public decimal ProfitThreshold { get => _profitThreshold.Value; set => _profitThreshold.Value = value; }
	public decimal AbsoluteFixation { get => _absoluteFixation.Value; set => _absoluteFixation.Value = value; }
	public decimal StabilizationPoints { get => _stabilizationPoints.Value; set => _stabilizationPoints.Value = value; }
	public int ExpirationMinutes { get => _expirationMinutes.Value; set => _expirationMinutes.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public OrderStabilizationStrategy()
	{
		_orderVolume = Param(nameof(OrderVolume), 0.1m).SetGreaterThanZero();
		_orderDistancePoints = Param(nameof(OrderDistancePoints), 20m).SetGreaterThanZero();
		_profitThreshold = Param(nameof(ProfitThreshold), -2m);
		_absoluteFixation = Param(nameof(AbsoluteFixation), 30m).SetNotNegative();
		_stabilizationPoints = Param(nameof(StabilizationPoints), 25m).SetGreaterThanZero();
		_expirationMinutes = Param(nameof(ExpirationMinutes), 20).SetNotNegative();
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame());
	}

	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		=> [(Security, CandleType)];

	protected override void OnReseted()
	{
		base.OnReseted();
		ClearPending();
		_entryPrice = 0m;
		_previousBody = 0m;
		_hasPreviousBody = false;
		_entryCandleTime = null;
	}

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);
		SubscribeCandles(CandleType).Bind(ProcessCandle).Start();
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		if (candle.State != CandleStates.Finished)
			return;

		var point = Security?.PriceStep ?? 1m;
		if (point <= 0m)
			point = 1m;

		var body = Math.Abs(candle.ClosePrice - candle.OpenPrice);
		var bodyLimit = StabilizationPoints * point;

		if (Position != 0 && (_entryCandleTime is null || candle.OpenTime > _entryCandleTime.Value))
		{
			var pnl = FloatingPnL(candle.ClosePrice, point);
			var oneSmall = body <= bodyLimit;
			var twoSmall = oneSmall && _hasPreviousBody && _previousBody <= bodyLimit;

			if ((oneSmall && pnl > ProfitThreshold) ||
				twoSmall ||
				(AbsoluteFixation > 0m && pnl >= AbsoluteFixation))
			{
				Flatten();
				ClearPending();
				_previousBody = body;
				_hasPreviousBody = true;
				return;
			}
		}

		if (_createdAt is DateTimeOffset created &&
			ExpirationMinutes > 0 &&
			candle.OpenTime - created >= TimeSpan.FromMinutes(ExpirationMinutes))
		{
			ClearPending();
			CreatePending(candle.ClosePrice, candle.OpenTime, point);
			_previousBody = body;
			_hasPreviousBody = true;
			return;
		}

		// Stops created from a finished candle can only be triggered by a later candle.
		if (_buyStop is null && _sellStop is null)
		{
			CreatePending(candle.ClosePrice, candle.OpenTime, point);
			_previousBody = body;
			_hasPreviousBody = true;
			return;
		}

		var hitBuy = _buyStop is decimal buy && candle.HighPrice >= buy;
		var hitSell = _sellStop is decimal sell && candle.LowPrice <= sell;

		if (hitBuy || hitSell)
		{
			var side = hitBuy && hitSell
				? (candle.ClosePrice >= candle.OpenPrice ? Sides.Buy : Sides.Sell)
				: hitBuy ? Sides.Buy : Sides.Sell;

			var trigger = side == Sides.Buy ? _buyStop!.Value : _sellStop!.Value;

			if (side == Sides.Buy)
			{
				BuyMarket(OrderVolume);
				_buyStop = null;
			}
			else
			{
				SellMarket(OrderVolume);
				_sellStop = null;
			}

			_entryPrice = trigger;
			_entryCandleTime = candle.OpenTime;
		}

		_previousBody = body;
		_hasPreviousBody = true;
	}

	private void CreatePending(decimal center, DateTimeOffset time, decimal point)
	{
		var distance = OrderDistancePoints * point;
		_buyStop = center + distance;
		_sellStop = center - distance;
		_createdAt = time;
	}

	private decimal FloatingPnL(decimal price, decimal point)
	{
		if (Position == 0m || _entryPrice == 0m)
			return 0m;

		var direction = Position > 0m ? 1m : -1m;
		var move = (price - _entryPrice) * direction;
		var stepPrice = Security?.StepPrice ?? 0m;

		return stepPrice > 0m
			? move / point * stepPrice * Math.Abs(Position)
			: move * (Security?.Multiplier ?? 1m) * Math.Abs(Position);
	}

	private void Flatten()
	{
		if (Position > 0m)
			SellMarket(Math.Abs(Position));
		else if (Position < 0m)
			BuyMarket(Math.Abs(Position));

		_entryPrice = 0m;
		_entryCandleTime = null;
	}

	private void ClearPending()
	{
		_buyStop = null;
		_sellStop = null;
		_createdAt = null;
	}
}
